---
title: "PospSigDB Documentation"
author: "Stefanos Apostle"
date: "2/23/2021"
output: html_document
---

> This file contains examples of how to both read and write .gmt files using the PospSigDB R-package functions to both use and contribute to the compilation of genesets in /secondary/projects/pospisilik/PospSigDB. This private lab database is meant to serve as a collective of gene sets from expiriments done in the lab in a centralized and secure location. More gene sets are available in the .gmt format at http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp and can be read into R using the load_gmt() function.

> These gene sets are formatted following the guidelines by MSigDB and shoulod be followed so that these same gene sets can be used by all GSEA packages, functions, and software. Format guidelines can be found at https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29. 

> These gene sets will also be organized in the following manner; first by the experimental data that these genes were derived from (RNA-seq, ChIP-seq, etc), then by project leader name, then finally by the project it is from. 


## To install the PospSigDB R-package, execute the following
```{r, eval = F}
library(devtools)
devtools::install_github("Stefanos-Apostle/PospSigDB")
```


## If installation of package from github not working, simply copy and paste the following functions into your global environment
```{r}
## Function to load in tab delim gene sets from .gmt formatted file
load_gmt <- function(tab_delim_gmt) {
  set <- read_delim(tab_delim_gmt, delim = "\n", col_names = FALSE)
  gmt_geneset_list = list()
  unlisted <- c()
  i = 1
  while (i <= length(set[[1]])){
    x = set[[1]][i]
    y = strsplit(x, "\t")
    l = length(y[[1]])
    sets = y[[1]][3:l]
    unlisted <- c(unlisted, sets)
    name = y[[1]][1]
    gmt_geneset_list[[name]] = sets
    i = i + 1
  }
  return(gmt_geneset_list)
}


## Function to write .gmt files from gene sets to be moved to /secondary/projects/pospisilik/PospSigDB/
write_gmt <- function(title, description, genes, path=".") {
  nt <- toupper(title)
  nt <- strsplit(nt, split = " ")[[1]]
  nt <- paste(nt, collapse = "_")
  contents <- c(nt, paste(">", description, sep = " "), genes)
  write.table(x = t(as.data.frame(contents)), file = paste(paste(path, "/", sep=''), nt, ".gmt", sep = ""), sep = "\t",
              row.names = F, col.names = F, quote=F)
}


```

## To Read the genes in a .gmt file; see the example utilizing load_gmt() below.

> This will read the .gmt file into R which can be stored in a variable. Since .gmt files can contain more than one gene set as seen below, each gene set is stored as an independant list object with the title as the list name. The same is done with a .gmt file, the list will just be a length of 1. 

```{r}
## Provide the path to the .gmt file and the function will read the genes into R
library(readr)
load_gmt("./IGN_genesets.gmt")
```



## To write a list of genes to a .gmt file for contribution to lab database; see example utilizing write_gmt() below.

> This can be used once an analysis has been done and for example found some genes of interest in the context of our project which we would like to store as a gene set for cross-analysis later on. For this, we must 

```{r}
## Specify the title of the gene set
title = "Upregulated Genes in Knockout"

## Describe the relevance of these genes
description = "These genes were found to be upregulated in our condition X."

## Genes of interest should be characters and in a vector.
genes <- c("GeneA", "GeneB", "GeneC", "GeneD", "GeneE")

## Finally specify the pasth to where you want this geneset to be written to in the PospSigDB
path = "." ## default will be written in the current working directory

## Finally execute the write_gmt function to properly format gene set
write_gmt(title = title, description = description, genes = genes, path = path)
```


> We can then see this gene set was made correctly when reading it back into R, which we can also see an example of a .gmt file which is read into R.

```{r}
## Provide the path to the .gmt file and the function will read the genes into R
load_gmt("./UPREGULATED_GENES_IN_KNOCKOUT.gmt")
```




































